File: htrace-core/src/go/src/org/apache/htrace/common/span.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package common

import (
	"encoding/json"
	"fmt"
	"strconv"
)

//
// Represents a trace span.
//
// Compatibility notes:
// When converting to JSON, we store the 64-bit numbers as hexadecimal strings rather than as
// integers.  This is because JavaScript lacks the ability to handle 64-bit integers.  Numbers above
// about 55 bits will be rounded by Javascript.  Since the Javascript UI is a primary consumer of
// this JSON data, we have to simply pass it as a string.
//

type TraceInfoMap map[string][]byte

type TimelineAnnotation struct {
	Time int64  `json:"time,string"`
	Msg  string `json:"msg"`
}

type SpanId int64

func (id SpanId) String() string {
	return fmt.Sprintf("%08x", id)
}

func (id SpanId) Val() int64 {
	return int64(id)
}

func (id SpanId) MarshalJSON() ([]byte, error) {
	return []byte(`"` + fmt.Sprintf("%016x", uint64(id)) + `"`), nil
}

func (id SpanId) UnMarshalJSON() ([]byte, error) {
	return []byte(`"` + strconv.FormatUint(uint64(id), 16) + `"`), nil
}

type SpanData struct {
	Begin               int64                `json:"b,string"`
	End                 int64                `json:"e,string"`
	Description         string               `json:"d"`
	TraceId             SpanId               `json:"i"`
	Parents             []SpanId             `json:"p"`
	Info                TraceInfoMap         `json:"n,omitempty"`
	ProcessId           string               `json:"r"`
	TimelineAnnotations []TimelineAnnotation `json:"t,omitempty"`
}

type Span struct {
	Id SpanId `json:"s,string"`
	SpanData
}

func (span *Span) ToJson() []byte {
	jbytes, err := json.Marshal(*span)
	if err != nil {
		panic(err)
	}
	return jbytes
}

File: htrace-htraced/src/web/app/models/span.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Span model
app.Span = Backbone.Model.extend({
  "defaults": {
    "spanId": null,
    "traceId": null,
    "processId": null,
    "parents": null,
    "description": null,
    "beginTime": 0,
    "stopTime": 0
  },

  shorthand: {
    "s": "spanId",
    "b": "beginTime",
    "e": "stopTime",
    "d": "description",
    "r": "processId",
    "p": "parents",
    "i": "traceId"
  },

  parse: function(response, options) {
    var attrs = {};
    var $this = this;
    $.each(response, function(key, value) {
      attrs[(key in $this.shorthand) ? $this.shorthand[key] : key] = value;
    });
    return attrs;
  },

  duration: function() {
    return this.get('stopTime') - this.get('beginTime');
  }
});

app.Spans = Backbone.PageableCollection.extend({
  model: app.Span,
  mode: "infinite",
  url: "/query",
  state: {
    pageSize: 10,
    lastSpanId: null,
    finished: false,
    predicates: []
  },
  queryParams: {
    totalPages: null,
    totalRecords: null,
    firstPage: null,
    lastPage: null,
    currentPage: null,
    pageSize: null,
    sortKey: null,
    order: null,
    directions: null,

    /**
     * Query parameter for htraced.
     */
    query: function() {
      var predicates = this.state.predicates.slice(0);
      var lastSpanId = this.state.lastSpanId;

      /**
       * Use last pulled span ID to paginate.
       * The htraced API works such that order is defined by the first predicate.
       * Adding a predicate to the end of the predicates list won't change the order.
       * Providing the predicate on spanid will filter all previous spanids.
       */
      if (lastSpanId) {
        predicates.push({
          "op": "gt",
          "field": "spanid",
          "val": lastSpanId
        });
      }

      return JSON.stringify({
        lim: this.state.pageSize + 1,
        pred: predicates
      });
    }
  },

  initialize: function() {
    this.on("reset", function(collection, response, options) {
      if (response.length == 0) {
        delete this.links[this.state.currentPage];
        this.getPreviousPage();
      }
    }, this);
  },

  parseLinks: function(resp, xhr) {
    this.state.finished = resp.length <= this.state.pageSize;

    if (this.state.finished) {
      this.state.lastSpanId = null;
    } else {
      this.state.lastSpanId = resp[this.state.pageSize - 1].s;
    }

    if (this.state.finished) {
      return {};
    }

    return {
      "next": "/query?query=" + this.queryParams.query.call(this)
    };
  },

  parseRecords: function(resp) {
    return resp.slice(0, 10);
  },

  setPredicates: function(predicates) {
    if (!$.isArray(predicates)) {
      console.error("predicates should be an array");
      return;
    }

    this.state.predicates = predicates;
  }
});
